src/body.rs

```diff
@@ -237,36 +237,42 @@ impl Sender {
                 return Ok(().into());
             }
 
-            try_ready!(tx
-                .as_mut()
-                .expect("tx only taken on error")
-                .poll_ready()
-                .map_err(::error::from));
-
-            if buf.remaining_mut() == 0 {
-                buf.reserve(8192);
-            }
+            if buf.len() == 0 {
+                if buf.remaining_mut() == 0 {
+                    buf.reserve(8192);
+                }
 
-            match body.read(unsafe { buf.bytes_mut() }) {
-                Ok(0) => {
-                    return Ok(().into())
-                },
-                Ok(n) => {
-                    unsafe { buf.advance_mut(n); }
-                    written += n as u64;
-                    let tx = tx.as_mut().expect("tx only taken on error");
-                    if let Err(_) = tx.send_data(buf.take().freeze().into()) {
-                        return Err(::error::timedout(None));
+                match body.read(unsafe { buf.bytes_mut() }) {
+                    Ok(n) => {
+                        if n > 0 {
+                            unsafe { buf.advance_mut(n); }
+                        }
+                    }
+                    Err(e) => {
+                        let ret = io::Error::new(e.kind(), e.to_string());
+                        tx
+                            .take()
+                            .expect("tx only taken on error")
+                            .abort();
+                        return Err(::error::from(ret));
                     }
                 }
-                Err(e) => {
-                    let ret = io::Error::new(e.kind(), e.to_string());
-                    tx
-                        .take()
-                        .expect("tx only taken on error")
-                        .abort();
-                    return Err(::error::from(ret));
+            }
+
+            if buf.len() > 0 {
```

Comment:
If the previous [match arm](https://github.com/seanmonstar/reqwest/pull/479/files#diff-14e8fe89f8eea2bff803db41c7dd66a5L251) is kept, I believe this `if` can be removed.

Reply:
No, it is the return condition.

Reply:
What I mean is, since there is already a check for `if buf.len() == 0` above, the only way *this* check would be `> 0` is if the `body.read` returned `Ok(0)`, right?

Reply:
I'm not sure I understand that. The best would be to look at the final code structure:

```
if len() == 0 { // buffer is empty
    // Try reading data into the buffer
}
if len() > 0 { // buffer was filled by this iteration or another
    // Try writing data to the transmission channel
}

Also note that `try_ready` macro returns early if the channel is not ready, in which case the buffer is filled and transmitted in different iterations.

Reply:
Sorry for not being clear, I'll try as you did with some pseudo code:

```rust
if len() == 0 {
    match body.read(but) {
        Ok(0) => return,
        Ok(n) => (),
        Err(e) => return e;
    }
}
// Only way to get here is if len was greater
// than 0, or read more than 0
```

Reply:
Oh right. If you really want that., Give me a minute or two...

Reply:
Done. I checked the real life case and it's OK.